### ISSUE TITLE
[FIRRTL] SFC Invalid Interpretation

### ISSUE THREAD

**Issue by seldridge** [llvm/circt#1494]
The FIRRTL spec treats invalid as undefined.  However, the Scala FIRRTL Compiler has a context-sensitive definition of invalid: either (1) constant zero or (2) undefined. This has the huge potential for implementation mismatches between the Scala FIRRTL Compiler (SFC) and CIRCT.

Currently, the SFC will typically optimize invalids that hit primops as constant zero.  E.g., SFC will optimize the following to `b <= a`.

```scala
circuit Foo :
  module Foo :
    input a: UInt<1>
    output b: UInt<1>

    wire w: UInt<1>
    w is invalid

    b <= xor(a, w)
```

Conversely, a FIRRTL compiler could optimize this to `b is invalid` because `xor(_, invalid) -> invalid` (a boolean operator of an undefined value should itself be undefined).

In other situations, the SFC may treat invalid like undefined. Both `when` and `validif` interpret invalid in this way.

Either:
```scala
circuit DontCareMux :
  module DontCareMux :
    input clock : Clock
    input reset : UInt<1>
    output io : {flip en : UInt<1>, out : UInt<8>}

    node _T = validif(io.en, UInt<7>("h07b"))
    io.out <= _T
```

Or:

```scala
circuit DontCareMux :
  module DontCareMux :
    input clock : Clock
    input reset : UInt<1>
    output io : {flip en : UInt<1>, out : UInt<8>}

    wire _T: UInt<7>
    _T is invalid
    when io.en :
      _T <= UInt<7>("h07b")
    io.out <= _T
```

Will both compile, in the SFC, to `io.out <= UInt<7>("h07b")`.  However, if a `mux` construction is used, this will produce a default assignment of zero in the following, similar construction. (Arguably, the mux construction is an oversight and https://github.com/chipsalliance/firrtl/issues/2011 is a request to align `mux` with `when`/`validif`.):

```scala
    wire _T_1 : UInt<7>
    _T_1 is invalid
    node _T = mux(io.en, UInt<7>("h07b"), _T_1)
    io.out <= _T
```

There is a further example of weird behavior in https://github.com/chipsalliance/firrtl/issues/2011#issuecomment-749745556 which shows that certain optimizations are disallowed if there is a requirement that an invalid source always returns the same value.

tl;dr: the SFC will typically interpret invalid as zero, but will selectively interpret invalid as undefined. I think this can be all cleaned up with an early pass that converts invalids to zeros (when appropriate) and leaves other invalids around for undefined-style optimization. This can then be enabled with an `-sfc-compat` option (or something like that).
